### PR TITLE
fix: implemented method to add CFN resource via overrides in auth

### PR DESCRIPTION
**Pull request by akshbhu**
#### Description of changes
- implemented method to add cfn resource via overrides in auth
- update types to include addResource calls

#### Description of how you validated changes
- tested Manually 
- 
#### Checklist

- [x] PR description included
- [ ] `yarn test` passes
- [ ] Tests are [changed or added](https://github.com/aws-amplify/amplify-cli/blob/master/CONTRIBUTING.md#tests)
- [ ] Relevant documentation is changed or added (and PR referenced)
- [ ] New AWS SDK calls or CloudFormation actions have been added to relevant test and service IAM policies

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/amplify-category-auth/src/provider-utils/awscloudformation/auth-stack-builder/auth-cognito-stack-builder.ts [hunk 1]
```diff
@@ -103,13 +104,25 @@ export class AmplifyAuthCognitoStack extends cdk.Stack implements AmplifyAuthCog
     this.lambdaTriggerPermissions = {};
   }
   addCfnResource(props: cdk.CfnResourceProps, logicalId: string): void {
-    throw new Error('Method not implemented.');
+    if (!this._cfnResourceMap.has(logicalId)) {
+      this._cfnResourceMap.set(logicalId, new cdk.CfnResource(this, logicalId, props));
+    } else {
+      throw new Error(`Cfn Resource with LogicalId ${logicalId} doesnt exist`);
```
Comment (jhockett) [inline packages/amplify-category-auth/src/provider-utils/awscloudformation/auth-stack-builder/auth-cognito-stack-builder.ts:110]: I think the error message here should be:
```ts
    throw new Error(`Cfn Resource with LogicalId ${logicalId} already exists.`);
```

Similarly for the errors below

**Review by sachscode — APPROVED** [review]

**Comment by henobi** [comment]
Hi guys,

we are currently stuck to deploy changes on all of our stages. We will really appreciate to have this change in a release soon because it will fix that issue.

Thank you! 

Hendrik

**Comment by lorodialog** [comment]
Hi @jhockett,
We really appreciate your support! ✌️
Please, would you be so kind and approve the changes @akshbhu implemented?
We really need this productive...
Thank you!

File: packages/amplify-category-auth/src/provider-utils/awscloudformation/auth-stack-builder/auth-cognito-stack-builder.ts (hunk 1 above)
Comment (henobi) [reply packages/amplify-category-auth/src/provider-utils/awscloudformation/auth-stack-builder/auth-cognito-stack-builder.ts]: Helo @jhockett, could you please review again. I think it is resolved now.
We really need that fix in amplify cli. Thank you!!

**Comment by lorodialog** [comment]
Hi @akshbhu,
now it is possible to add a custom resource via e.g.
```
/* Login with Apple */
  resources.addCfnResource(
    {
      type: 'AWS::Cognito::UserPoolIdentityProvider',
      properties: {
        UserPoolId: { Ref: 'UserPool' },
        ProviderName: 'SignInWithApple',
        ProviderType: 'SignInWithApple',
```
But how can we define dependencies?
```
resources.userPoolClientWeb.addDependsOn(???);
```
Thank you so much!
Best regards!

**Comment by henobi** [comment]
Hi @akshbhu, 
I think `resources.addCfnResource(...)` should return the CDK resource instead of void.

Best, Hendrik

**Comment by akshbhu** [comment]
Hi @henobi  @lorodialog 

You are actually correct . We are working to expose complete interface with all get/add CFN resources/parameters.

For the workaround ,can you try this :

``` typescript
  resources.addCfnResource(
    {
      type: 'AWS::Cognito::UserPoolIdentityProvider',
      properties: {
        UserPoolId: { Ref: 'UserPool' },
        ProviderName: 'SignInWithApple',
        ProviderType: 'SignInWithApple',
}, 'resourcelogicalid')

// Now a way to access this resources :
// get the root node in cdk object through userPool
    const rootNode = resources.userPool.node.root.node;
// try adding this resource in the end of override file  so that we know the index
    const Customresource = rootNode.findAll()[rootNode.findAll().length-1] as CfnUserPoolDomain;
// add dependency
    resources.userPoolClientWeb.addDependsOn(Customresource);
```

this dependency will be like this in CFN:
```json
    "UserPoolClientWeb": {
      "Type": "AWS::Cognito::UserPoolClient",
      "Properties": {
        "UserPoolId": {
          "Ref": "UserPool"
        },
        "ClientName": "i9240bd1db5c4_app_clientWeb",
        "RefreshTokenValidity": {
          "Ref": "userpoolClientRefreshTokenValidity"
        }
      },
      "DependsOn": [
        "resourcelogicalid",
        "UserPool"
      ]
    },
```

**Pull request merged**